Added support for multisort (#378) - refactored

ojninja16 opened the pull request:
This PR adds support for **multi-sort functionality** in the API
### Changes Made
   - Simplified the code as suggested.
   - Added support for reusing the last item in shorter arrays.

## Summary by Sourcery

Adds support for multi-sort functionality to the API, allowing users to sort results by multiple fields, orders, and modes.

Enhancements:
- Improves sorting functionality by allowing multiple fields, sort orders, and sort modes to be specified.
- Simplifies code and adds support for reusing the last item in shorter arrays.

ojninja16 commented:
hi @rjchallis I have made the suggested changes and simplified the implementation, please let me know if something else needs to be modified.

rjchallis commented:
Thanks for your efforts on this. I've just been trying it out and ran into a couple of issues
- as you still have the enums in `api-v2.yaml`, the API is rejecting `sortOrder=asc%2desc`. Commenting out the `sortOrder` and `sortMode` enums gets past that error, but that means that validation will have to move into `setSortBy` to make sure the values are allowed.
- The other issue is that `sortBy`, `sortOrder` and `sortMode` are being passed to `setSortBy` as arrays because of `type: array` in the API schema. So when checking if the value is array or string to see if it needs splitting on comma, what actually gets checked is, e.g. `[ "asc,desc" ]`, changing the type back to string should fix this.

For note, I've been trying it out on the test data with:
```
curl "http://localhost:3000/api/v2/search?result=taxon&includeEstimates=true&taxonomy=ncbi&query=tax_tree%28canidae%29&size=10&sortBy=assembly_level%2Cassembly_span&sortOrder=asc%2Cdesc&sortMode=max%2Cmin&fields=assembly_level%2Cassembly_span" -H "accept: text/tab-separated-values"
```

ojninja16 commented:
HI @rjchallis  Thanks for your feedback! I’m a bit confused, I might be misunderstanding something about the direction and want to clarify before making further changes.

### Initial Requirement:
You initially asked for `setSortBy` to:
1. Split inputs on `,` delimiter **or** accept array parameters.
2. Return an array if there’s a list of values.

Based on this, I updated `api-v2.yaml` to use `type: array` with `items: string` 

### Current Issues:
You pointed out that:
1. The enums in `api-v2.yaml` are causing the API to reject comma-separated values (e.g., `sortOrder=asc,desc`).
2. `type: array` is causing `sortBy`, `sortOrder`, and `sortMode` to be passed as arrays (e.g., `["asc,desc"]`), which breaks the logic in `setSortBy`.

### Questions:
1. Should we change the schema back to `type: string` to avoid the issue of arrays being wrapped, even though the initial requirement was to support arrays?
2. If we use `type: string`, how should we handle **explicit arrays** (e.g., `sortBy[]=field1&sortBy[]=field2`), which were part of the initial requirement or should we be using `oneOf` in `api-v2.yaml` to  allow `sortBy`, `sortOrder`, and `sortMode` to be either a `string` or an `array`

I just want to make sure I’m aligned with the intended approach before proceeding. Thanks for your help!

rjchallis commented:
Sorry for the confusion! I was initially thinking that it would be good to support comma delimited string or arrays for more flexibility, but on reflection the values are simple enough that it seems reasonable to drop the idea of supporting explicit arrays so there is just one input type. Explicit array syntax is less user friendly so if only one is supported, I would prefer it to be the string.

In my last reply I was considering suggesting a combination of flatMap and split to handle possible arrays of comma separated strings, but then it would be possible for a user to pass 5 parameters as `["asc,desc", "desc", "desc,asc"]`, which doesn't seem like it should be supported and adding validation to prevent this while still allowing comma separated strings would just add complexity.

ojninja16 commented:
got it, thanks for the clarification! I will make the changes and raise a PR soon😄

ojninja16 commented:
hi @rjchallis I have refactored the setSortby.js and made changes in api-v2.yml as per the suggestions , let me know if there are any issues

ojninja16 commented:
I ran Prettier and pushed the formatted code, it should be good now
